econml/utilities.py

```diff
@@ -1391,6 +1393,91 @@ def predict(self, X):
         return self._clf.predict_proba(X)[:, 1:]
 
 
+class _TransformerWrapper:
+    """Wrapper that takes a featurizer as input and adds jacobian calculation functionality"""
+
+    def __init__(self, featurizer):
+        pass
+        self.featurizer = featurizer
+
+    def fit(self, X):
+        return self.featurizer.fit(X)
+
+    def transform(self, X):
+        return self.featurizer.transform(X)
+
+    def fit_transform(self, X):
+        return self.featurizer.fit_transform(X)
+
+    def get_feature_names(self, X):
+        if hasattr(self.featurizer, 'get_feature_names'):
+            return self.featurizer.get_feature_names(X)
+
+        return []
+
+    def jac(self, X):
+        if hasattr(self.featurizer, 'jac'):
+            return self.featurizer.jac(X)
+        elif (isinstance(self.featurizer, PolynomialFeatures)):
+            self.featurizer.fit(X)
+            powers = self.featurizer.powers_
+            result = np.zeros(X.shape + (self.featurizer.n_output_features_,))
+            for i in range(X.shape[1]):
+                p = powers.copy()
+                c = powers[:, i]
+                p[:, i] -= 1
+                M = np.float_power(X[:, np.newaxis, :], p[np.newaxis, :, :])
+                result[:, i, :] = c[np.newaxis, :] * np.prod(M, axis=-1)
+            return result
+
+        else:
+
+            squeeze = []
+
+            n = X.shape[0]
+
+            d_t = X.shape[-1] if ndim(X) > 1 else 1
+
+            X_out = self.transform(X)
+            d_f_t = X_out.shape[-1] if ndim(X_out) > 1 else 1
+
+            epsilon = 0.001
+
+            jacob = np.zeros((n, d_t, d_f_t))
+
+            if ndim(X) == 1:
+                squeeze.append(1)
+                X = X[:, np.newaxis]
+            if ndim(X_out) == 1:
+                squeeze.append(2)
+
+            # for every dimension of the treatment add some epsilon and observe change in featurized treatment
+            for k in range(d_t):
+                eps_matrix = np.zeros(shape=X.shape)
+                eps_matrix[:, k] = epsilon
+
+                X_in_plus = X + eps_matrix
+                X_in_plus = X_in_plus.squeeze(axis=1) if 1 in squeeze else X_in_plus
+                X_out_plus = self.transform(X_in_plus)
+                X_out_plus = X_out_plus[:, np.newaxis] if 2 in squeeze else X_out_plus
+
+                X_in_minus = X - eps_matrix
+                X_in_minus = X_in_minus.squeeze(axis=1) if 1 in squeeze else X_in_minus
+                X_out_minus = self.transform(X_in_minus)
+                X_out_minus = X_out_minus[:, np.newaxis] if 2 in squeeze else X_out_minus
+
+                diff = X_out_plus - X_out_minus
```

Comment:
From a numerical computation point of view, do you know if there is a reason to prefer this symmetric approach (f(x+e)-f(x-e))/2e compared to the non-symmetric approach (f(x+e)-f(x))/e, which would avoid the second call to transform?

Reply:
It is my shallow understanding that central difference methods are less error-prone than forward/backward difference methods. There may be nuance to that statement but I believe in general that is about true.

Interesting link:
https://math.stackexchange.com/questions/2120946/why-is-central-difference-preferred-over-backward-and-forward-difference-in-conv

Reply:
Thanks, makes sense.